core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java

```diff
@@ -1109,36 +1054,139 @@ private boolean shouldPerformTransactionOperation(StoreStatus storeStatus, Predi
+                     }).last(0L);
 
             },
             this::releaseReadLock
       ).toCompletionStage();
    }
 
-   private <K, V> MVCCEntry<K, V> acquireKeyFromContext(InvocationContext ctx, WriteCommand command, Object key,
-         BiPredicate<WriteCommand, Object> commandKeyPredicate) {
-      if (commandKeyPredicate.test(command, key)) {
+   <K> Flowable<NonBlockingStore.SegmentedPublisher<K>> createRemoveFlowable(Flowable<K> keyRemoveFlowable,
+         boolean shared, boolean segmented, StoreStatus storeStatus) {
+      Flowable<NonBlockingStore.SegmentedPublisher<K>> removeFlowable;
+      if (segmented) {
+         removeFlowable = keyRemoveFlowable
+               .groupBy(keyPartitioner::getSegment)
+               .map(SegmentPublisherWrapper::wrap);
+         if (shared) {
+            removeFlowable = filterSharedSegments(removeFlowable);
+         }
+      } else {
+         if (shared) {
+            keyRemoveFlowable = keyRemoveFlowable.filter(k ->
+                  distributionManager.getCacheTopology().getDistribution(k).isPrimary());
+         }
+         removeFlowable = Flowable.just(SingleSegmentPublisher.singleSegment(keyRemoveFlowable));
+      }
+
+      if (trace) {
+         removeFlowable = removeFlowable.doOnSubscribe(sub ->
+               log.tracef("Store %s has subscribed to remove batch", storeStatus.store));
+         removeFlowable = removeFlowable.map(sp -> {
+            int segment = sp.getSegment();
+            return SingleSegmentPublisher.singleSegment(segment, Flowable.fromPublisher(sp)
+                  .doOnNext(keyToRemove -> log.tracef("Emitting key %s for removal from segment %s",
+                        keyToRemove, segment)));
+         });
+      }
+
+      return removeFlowable;
+   }
+
+   <K, V> Flowable<NonBlockingStore.SegmentedPublisher<MarshallableEntry<K, V>>> createWriteFlowable(
+         Flowable<MarshallableEntry<K, V>> entryWriteFlowable, boolean shared, boolean segmented,
+         AtomicLong writeCount, StoreStatus storeStatus) {
+      Flowable<NonBlockingStore.SegmentedPublisher<MarshallableEntry<K, V>>> writeFlowable;
+      if (segmented) {
+         // Note the writeCount includes entries that aren't written due to being shared
+         // at this point
+         entryWriteFlowable = entryWriteFlowable.doOnNext(obj -> writeCount.incrementAndGet());
+         writeFlowable = entryWriteFlowable
+               .groupBy(me -> keyPartitioner.getSegment(me.getKey()))
+               .map(SegmentPublisherWrapper::wrap);
+         if (shared) {
+            // The writeCount will be decremented for each grouping of values ignored
+            writeFlowable = filterSharedSegments(writeFlowable);
+         }
+      } else {
+         if (shared) {
+            entryWriteFlowable = entryWriteFlowable.filter(me ->
+                  distributionManager.getCacheTopology().getDistribution(me.getKey()).isPrimary());
+         }
```

Comment:
Is there a real cost to changing this to:

``java
else {
         entryWriteFlowable = entryWriteFlowable
               .filter(me -> !shared || distributionManager.getCacheTopology().getDistribution(me.getKey()).isPrimary())
               .doOnNext(obj -> writeCount.incrementAndGet());
         writeFlowable = Flowable.just(SingleSegmentPublisher.singleSegment(entryWriteFlowable));
      }
```

The same approach can be taken with `createRemoveFlowable`.

Reply:
Hard to say, you have the extra lambda allocation and will have to invoke it for every entry in the write or remove. I would say we just leave it how it is personally.